src/runtimes/wapc.rs

```diff
@@ -284,13 +369,57 @@ impl<'a> Runtime<'a> {
             }
             Err(e) => {
                 error!(error = e.to_string().as_str(), "waPC communication error");
+                if e.to_string()
+                    .as_str()
+                    .contains(WAPC_EPOCH_INTERRUPTION_ERR_MSG)
+                {
+                    // TL;DR: after code execution is interrupted because of an
```

Comment:
questions:

we can't use [`epoch_deadline_async_yield_and_update()`](https://docs.rs/wasmtime/latest/wasmtime/struct.Store.html#method.epoch_deadline_async_yield_and_update) because we aren't async (we spawn different threads and each has a synchronous wasmtime::Engine). Correct?

We can't use [`epoch_deadline_callback()`](https://docs.rs/wasmtime/latest/wasmtime/struct.Store.html#method.epoch_deadline_callback) to yield to other policy and bump the deadline, because we only have 1 policy per wasmtime::Engine, each in their own threads (good for containerization). The policy must complete evaluation before deadline, there's no yield and resume. Correct?

Reply:
> We don't want yield and resume of policy evaluation, because that would block the k8s resource in the admissioncontroller anyways. Correct?

We don't want to resume because we don't know if the code will ever complete its execution. After a certain amount of time the Kubernetes API server will declare the response to be a timeout. IMHO I think it's useless to keep giving control back to the wasm code hoping it completes its execution, just to end up not having anybody interested in the response (because of the timeout).

> we can't use epoch_deadline_async_yield_and_update() anyways because we aren't async (we spawn different threads in the worker pool and each has a synchronous wasmtime::Engine). Correct?

True, this is can be used only when the async feature is enabled inside of wasmtime and all the interactions with wasmtime are done via the async methods. This would require quite a big set of changes to waPC. This is something I want to contribute, but not right now.

However, I don't want to give control back to the wasm guest once the epoch interruption took place, see above.

> We can't use epoch_deadline_callback() to yield to other policy and bump the deadline, because we only have 1 policy per wasmtime::Engine, each in their own threads (good for containerization). Correct?

We could use the [Store::epoch_deadline_callback](https://docs.rs/wasmtime/4.0.0/wasmtime/struct.Store.html#method.epoch_deadline_callback) to allow the wasm code to run longer, all without having to enable async support inside of wasmtime. However, I don't want to give control back to the wasm guest once the epoch interruption took place, see above.

However, if we were to give control back, we would still need to abort the wasm guest code after some time because otherwise we would end up having one of our worker threads stuck evaluating the same request.
Other similar requests could be made against policy server, leading to all the workers being stuck in an endless evalution, leading to DOS attack against policy server. We know that would have bad effects on the Kubernetes API server and the cluster itself.

Reply:
Thanks for the detailed answers, resolving!